A couple of enhancements to user-visible threading semantics

RandomShaper opened the pull request:
Kept in two commits because, regardless both related to threading, the changes are heavily unrelated otherwise.

I'd appreciate especially careful reviewing of my `GDVIRTUAL` changes, as this is the first time I deal with that.

RandomShaper committed:
Let read thread guards allow access to out-of-tree nodes

RandomShaper commented:
CC @SlugFiller

SlugFiller commented:
Cool! Nice to see things are happening quickly. Here's hoping it gets merged soon.

YuriSizov commented:
> Here's hoping it gets merged soon.

Did you test it, do you approve it? 🙃

SlugFiller commented:
With just the first commit, I hit the same deadlock in running `ResourceSaver.save` in parallel.

With both commits, and `_can_import_threaded` set to return `false`, it works perfectly. However, if I don't implement `_can_import_threaded`, Godot poofs on import. So something is definitely not right about the `GDVIRTUAL_CALL`.

RandomShaper commented:
@SlugFiller, thank you for the feedback. I'm wondering, what happens if you override but return `true`?

SlugFiller commented:
When returning `true` it works with single, or deadlocks on `ResourceSaver.save` if done in parallel. If I don't override it, it poofs even on single.

RandomShaper committed:
Allow EditorImportPlugin to override can_import_threaded()

RandomShaper commented:
I've fixed a very silly mistake causing infinite recursion if the method is not overridden. Please test again with `false`. The problem with parallel saving will have to be investigated and fixed separately as it's not in the scope of this PR.

SlugFiller commented:
Poofing is gone, and returning `false` works for importing 3 scenes at once. Seems great.

akien-mga commented on doc/classes/EditorImportPlugin.xml:
```diff
@@ -120,6 +120,13 @@
+			<return type="bool" />
+			<description>
+				Tells whether this importer can be run in parallel on threads, or, on the contrary, it's only safe for the editor to call it from the main thread, for one file at a time.
+				If this method it not overridden, it will return [code]true[/code] by default (i.e., safe for parallel importing).
```
```suggestion
				If this method is not overridden, it will return [code]true[/code] by default (i.e., safe for parallel importing).
```
